### PR TITLE
Cluster analysis sub-chapter

### PR DESCRIPTION
**Pull request by damirpolat**
I added the first draft for clustering section. 
Please let me know if you have any comments or suggestions. 

Thank you,
Damir

### CODE REVIEW

File: bookdown/07-special-cluster.Rmd [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+Since clustering is a type of unsupervised learning, `TaskClust` is slightly different from `TaskRegr` and `TaskClassif` objects.
+More specifically:
+
+  * `truth()` function is missing because observations are not labeled.
+  * `target` field is empty and will return `character(0)` if accessed anyway.
+
+
+### Train and Predict
+Clustering learners provide both `train` and `predict` methods.
+The analysis typically consists of building clusters using all available data.
```
Comment (larskotthoff) [inline bookdown/07-special-cluster.Rmd:25]: Would mention explicitly that this refers to training here.

File: bookdown/07-special-cluster.Rmd [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+Nevertheless, some learners can assign new observations to existing groups with `predict`.
+However, prediction does not always make sense, e.g., for hierarchical clustering.
```
Comment (larskotthoff) [inline bookdown/07-special-cluster.Rmd:27]: Why does it not make sense?

File: bookdown/07-special-cluster.Rmd [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+Moreover, some learners estimate the probability of each observation belonging to a given cluster.
+`predict_types` field gives a list of prediction types for each learner.
+
+After training, the `model` field stores a learner's model that looks different for each learner depending on the underlying library.
+`predict` returns a `PredictionClust` object that gives a simplified view of the learned model.
+If the data given to the `predict` method is the same as the one on which the learner was trained, `predict` simply returns cluster assignments for the "training" observations.
+On the other hand, if the test set contains new data, `predict` will estimate cluster assignments for that data set (some learners do not do that, so read the documentation first).
```
Comment (larskotthoff) [inline bookdown/07-special-cluster.Rmd:34]: What happens if you try to do this with a learner that doesn't support it?

File: bookdown/07-special-cluster.Rmd [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+learner$model
+
+# make "predictions" for the same data
+prediction = learner$predict(task)
+autoplot(prediction, task)
+```
+
+
+### Measures
+To assess the quality of clustering, there are a few built-in evaluation metrics available.
```
Comment (larskotthoff) [inline bookdown/07-special-cluster.Rmd:62]: I would mention the fundamental difference to other types of problems here again -- no ground truth. Briefly mention how you can assess the quality of a clustering nevertheless.

File: bookdown/07-special-cluster.Rmd [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+One of them is silhouette quality index that measures how well each point belongs to its assigned cluster versus neighboring cluster.
+Silhouette values are in $[-1, 1]$ range.
+
+Points with silhouette closer to:
+
+   * 1 are well clustered
+   * 0 lie between two clusters
+   * -1 likely placed in the wrong cluster
+
+Using silhouette plots, one can visually assess the quality of the analysis and choose the most optimal number of clusters for the data.
```
Comment (larskotthoff) [inline bookdown/07-special-cluster.Rmd:72]: Not sure if "most optimal" is true here...

File: bookdown/07-special-cluster.Rmd [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+learner$train(task)
+learner$model
+
+# make "predictions" for the same data
+prediction = learner$predict(task)
+autoplot(prediction, task)
+```
+
+
+### Measures
```
Comment (larskotthoff) [inline bookdown/07-special-cluster.Rmd:61]: This section is a bit weird in that in only mentions one measure, which is only shown through visualization (the next section). I would mention at least one other measure here, and something that gives a numeric quantity, similar to e.g. accuracy in supervised problems.

File: bookdown/07-special-cluster.Rmd [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+   * 0 lie between two clusters
+   * -1 likely placed in the wrong cluster
+
+Using silhouette plots, one can visually assess the quality of the analysis and choose the most optimal number of clusters for the data.
+The red dotted line shows the mean silhouette value and each bar represents a data point.
+If most points in each cluster have an index higher than the mean silhouette, the number of clusters is chosen well.
+
+```{r 07-special-cluster-003, message=FALSE, warning=FALSE}
+# silhouette plot allows to visually inspect the quality of clustering
+autoplot(prediction, task, type = "sil")
```
Comment (larskotthoff) [inline bookdown/07-special-cluster.Rmd:78]: Presumably there's a way to get an average score as well, so you can assess this without a plot?

**Review by larskotthoff — CHANGES_REQUESTED** [review]

**Comment by damirpolat** [comment]
I addressed the comments. Please let me know if there's anything else I can improve.

File: bookdown/07-special-cluster.Rmd [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+
+### Measures
+The difference between supervised and unsupervised learning is that there is no ground truth data in unsupervised learning.
+In a supervised setting, such as classification, we would need to compare our predictions to true labels.
+Since clustering is an example of unsupervised learning, there are no true labels to which we can compare.
+However, we can still measure the quality of cluster assignments by quantifying how closely objects within the same cluster are related (cluster cohesion) as well as how distinct different clusters are from each other (cluster separation).
+
+To assess the quality of clustering, there are a few built-in evaluation metrics available.
+One of them is within sum of squares which quantifies cluster cohesion.
```
Comment (larskotthoff) [inline bookdown/07-special-cluster.Rmd:89]: Sum of squares of what? Distances?

**Review by larskotthoff — CHANGES_REQUESTED** [review]
Thanks, looks much better, just one tiny thing I noticed.

File: bookdown/07-special-cluster.Rmd (hunk 8 above)
Comment (damirpolat) [reply bookdown/07-special-cluster.Rmd]: I added more info on the measure. Thank you!

**Comment by larskotthoff** [comment]
Looks good to me -- would be great if somebody else could have a look at it as well!

**Comment by damirpolat** [comment]
New version of mlr3cluster is on cran and all checks are now passing.

**Comment by mllg** [comment]
Thanks Damir!

**Pull request merged**